### ISSUE TITLE
Inefficient exponentially decayed sum

### ISSUE THREAD

**Issue by deepers** [google/jax#3083]
I would like to write a JAX function that computes a running exponentially decayed sum along the zeroth axis. I am interested in being able to take the derivative with respect to the decay parameter.

I am using `index_update()` to accumulate values. According to the documentation, the JIT-compiled version should perform the operations in place. But what I find is that the JIT-compiled function runs even slower than the non-JIT compiled function, and doesn't even complete for large arrays of the size I would like to use in practice. Code and output are below.

Any ideas of how I can efficiently implement this function in JAX?
```
3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]
NumPy: 1.18.4
JAX: 0.1.66
Running on an array of shape (30, 30)
Numpy: 0:00:00.000104
JAX: 0:00:00.326483
JIT: 0:00:04.847453
```
```python
import datetime
import sys

import jax
import numpy
import numpy as np
from jax import numpy as jnp, ops, jit


def decay_sum_jax(x, decay):
    y = jnp.zeros(x.shape)
    for t in range(x.shape[0]):
        y = ops.index_update(y, t, x[t] + y[t - 1] * decay)
    return y


decay_sum_jit = jit(decay_sum_jax)


def decay_sum_numpy(x, decay):
    y = np.zeros(x.shape)
    for t in range(x.shape[0]):
        y[t] = x[t] + y[t - 1] * decay
    return y


if __name__ == '__main__':
    print(sys.version)
    print(f'NumPy: {str(numpy.__version__)}')
    print(f'JAX: {jax.__version__}')

    a = np.random.randn(30, 30)
    print(f'Running on an array of shape {a.shape}')

    t1 = datetime.datetime.now()
    decay_sum_numpy(a, 0.9)
    t2 = datetime.datetime.now()
    print(f'Numpy: {t2 - t1}')

    decay_sum_jax(a, 0.9)
    t3 = datetime.datetime.now()
    print(f'JAX: {t3 - t2}')

    decay_sum_jit(a, 0.9)
    t4 = datetime.datetime.now()
    print(f'JIT: {t4 - t3}')
```

**Comment by hawkinsp** [comment]
A few observations:
* You're mostly not measuring the execution time; you're measuring compilation time which happens the first time you execute a function and is then cached. You should use a warmup iteration if you want to measure execution time.
* If you want to measure execution time, please read: https://jax.readthedocs.io/en/latest/async_dispatch.html
You must use `.block_until_ready()` or consume the value from the host in some way.
* Your computation will be slow to compile because it unrolls the body of the `for` loop. One way to improve compilation time would be to use `lax.scan` (https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html#jax.lax.scan) which does pretty much exactly what you want here.

**Comment by jakevdp** [comment]
You might also get some improvement by re-expressing your function as a linear operator, rather than as an iterative procedure. This gives the same output as your original function:
```python
def decay_mat_jax(x, decay):
    N = x.shape[0]
    M = decay ** (jnp.arange(N)[:, jnp.newaxis] - jnp.arange(N))
    M = jnp.where(M <= 1, M, 0)
    return M @ x
```

**Comment by deepers** [comment]
Thanks @hawkinsp and @jakevdp for your suggestions! Using `scan()` plus `jit()` worked like a charm. Here is a comparison of the execution results on a 1000 x 3000 array (using `%timeit` and `.block_until_ready()`).

Numpy: 11.6 ms
Scan + JIT: 6.3 ms
Lin Op + JIT: 106 ms

One question about `jit()`. Is it ever necessary to call it multiple times, e.g. `jit(grad(jit(f)))`, or is it always fine to only put in on the outside?

For the record, here is the implementation of my function using `scan()`.

```python
def decay_sum_scan(x, decay):
    def f(sxtm1, xt):
        b = xt + decay * sxtm1
        return b, b

    return scan(f, jnp.zeros(x.shape[1:]), x)[1]
```

**Comment by hawkinsp** [comment]
At the moment, only the outermost `jit` has any effect. The inner ones do no harm but don't really affect anything. (They end up as `Call` operators in the HLO IR, which end up inlined I believe.)

Closing since it seems you are happy now!

**Comment by deepers** [comment]
Thanks again!
